packages/dicomImageLoader/src/imageLoader/internal/rangeRequest.ts

```diff
@@ -0,0 +1,241 @@
+import external from '../../externalModules';
+import { getOptions } from './options';
+import { LoaderXhrRequestError, LoaderXhrRequestPromise } from '../../types';
+import metaDataManager from '../wadors/metaDataManager';
+import extractMultipart from '../wadors/extractMultipart';
+import { LossyConfiguration } from 'core/src/types';
+
+const loadTracking: { [key: string]: { loaded: number; total: number } } = {};
+
+const streamCache: {
+  [key: string]: {
+    byteArray?: Uint8Array;
+    initialBytes: number;
+    totalRanges: number;
+    totalBytes?: number;
+    rangesFetched: number;
+  };
+} = {};
+
+/**
+ * Performs a range or thumbnail request.
+ * The configuration of exactly what is requested is based on the transfer
+ * syntax provided.
+ * Note this generates 1 response for each call, and those reponses may or may
+ * not be combined with each other depending on the configuration applied.
+ *
+ * * HTJ2K Streaming TSUID -> Use actual range requests, and set it up for partial
+ *   docding.
+ * * JLS and Non-streaming HTJ2K -> Use a thumbnail endpoint followed by normal
+ *   endpoint
+ *
+ * @param url - including an fsiz parameter
+ * @param imageId - to fetch for
+ * @param defaultHeaders  - to add to the request
+ * @returns Compressed image data
+ */
+export default function rangeRequest(
+  url: string,
+  imageId: string,
+  defaultHeaders: Record<string, string> = {},
+  retrieveOptions: LossyConfiguration = {}
+): LoaderXhrRequestPromise<{
+  contentType: string;
+  imageFrame: Uint8Array;
+  complete: boolean;
+  loadNextRange: () => any;
+}> {
+  const { cornerstone } = external;
+  const options = getOptions();
+
+  let initialBytes = options.initialBytes;
+  if (typeof initialBytes === 'function') {
+    const metaData = metaDataManager.get(imageId);
+    initialBytes = initialBytes(metaData, imageId);
+  }
+  if (!Number.isInteger(initialBytes)) {
+    throw new Error(
+      `initialBytes must be an integer or function that returns an integer.`
+    );
+  }
+
+  let totalRanges = options.totalRanges;
```

Comment:
give we have `rangesFetched`, then my suggestion is instead of `totalRanges` we use `totalRangesToFetch` or similar